### PR TITLE
`static_column_depth` interface and numerical bottom height in `AbstractGridFittedBottom`

### PR DESCRIPTION
**Pull request by simone-silvestri**
This PR introduces an interface for `static_column_depth ` that returns the height of the water column (not including the free surface). 
This is simply `grid.Lz` in regular grids.

For `AbstractGridFittedBottom` immersed grids, the column height is readily read from the immersed boundary, provided that the immersed boundary represents the _numerical_ bottom rather than the _physical_ bottom. 

Therefore, this PR changes the constructor of the  `ImmersedBoundaryGrid` to store the z-coordinate of the last immersed cell. In this way, the bottom height is also uniquely defined for `GridFittedBottom`, removing the necessity of having a`CenterImmersedCondition` and a `InterfaceImmersedCondition`. 

An immediate application of this interface is in the `SplitExplicitFreeSurface` where the bottom_height was previously stored.

### CODE REVIEW

**Comment by simone-silvestri** [comment]
Correction: `InterfaceImmersedCondition` and `CenterImmersedCondition` might still useful during construction because they inform where the bottom height will be.

**Comment by glwagner** [comment]
Is `column_height` a good name?

File: src/ImmersedBoundaries/grid_fitted_bottom.jl
Comment (simone-silvestri) [reply src/ImmersedBoundaries/grid_fitted_bottom.jl:24]: I though we want to refer to `z` as the coordinate and height as a positive value. 
At least, this is the direction we are taking in ClimaOcean where we are distinguishing very carefully between `z` and `h` or `height`. 
In this case, this is technically a coordinate, not a height, for example we have now
https://github.com/CliMA/Oceananigans.jl/blob/989108a6205bd39f7b215b2145ff5111015610df/src/ImmersedBoundaries/grid_fitted_bottom.jl#L109 which conflates the meaning of z with height. I thought we had decided this over at ClimaOcean and I wanted to update this code to reflect the decision we had made.

Comment (glwagner) [reply src/ImmersedBoundaries/grid_fitted_bottom.jl:24]: > I though we want to refer to z as the coordinate and height as a positive value.

Usually "z" is called "geopotential height". What would you call "z" instead? Altitude?

Comment (glwagner) [reply src/ImmersedBoundaries/grid_fitted_bottom.jl:24]: > this is technically a coordinate, not a height

Another name for z is "height coordinate".

Comment (jagoosw) [reply src/ImmersedBoundaries/grid_fitted_bottom.jl]: As someone who hasn't used immersed boundaries that much, I always find it a bit confusing that it's called bottom height because it makes me think I'm meant to put in a positive "height" from the bottom of the domain, but I would know that "z" is meant to be from the top

**Comment by simone-silvestri** [comment]
Shall we merge this as is without changing the name of the `bottom_height` then open an issue were we can discuss the name change?

**Comment by simone-silvestri** [comment]
There was another `total_depth` in the `TurbulenceClosures` module. I think the name `domain_depth` is appropriate (the domain is static) so I unified the two. 

There are also a couple of `depth` functions defined in the `TurbulenceClosures` module that are 3-dimensional. To distinguish from those, I believe `domain_depth` should accept two integers (i, j) instead of 3.

File: src/MultiRegion/multi_region_field.jl
Comment (glwagner) [inline src/MultiRegion/multi_region_field.jl:74]: `arch` should be an argument to `reconstruct_global_field` if we need to do this on CPU for some reason

Comment (simone-silvestri) [reply src/MultiRegion/multi_region_field.jl:74]: agreed

**Comment by glwagner** [comment]
I see `domain_depth`. What is the proposed interface?

**Comment by simone-silvestri** [comment]
I proposed `domain_depth` to distinguish it from the already existing `depth` interface that tracks the `depth` as a function of `k` which is used in the turbulence closure module. I think `domain_depth` communicates the idea of a total static depth.

**Comment by glwagner** [comment]
Ok we will call it `static_column_depth`

File: Project.toml
Comment (glwagner) [inline Project.toml:87]: Put this in another PR!!!!!

File: src/Grids/grid_utils.jl
Comment (glwagner) [inline src/Grids/grid_utils.jl:318]:
```suggestion
##### Static column depth
```

File: src/ImmersedBoundaries/grid_fitted_bottom.jl
Comment (glwagner) [inline src/ImmersedBoundaries/grid_fitted_bottom.jl:47]: indenting is messed up. Also why does this change?

File: src/ImmersedBoundaries/grid_fitted_bottom.jl
Comment (glwagner) [inline src/ImmersedBoundaries/grid_fitted_bottom.jl:97]: I think "correct" is too vague, can we go back to clamp or something more specific?

File: validation/multi_region/multi_region_near_global_quarter_degree.jl
Comment (glwagner) [inline validation/multi_region/multi_region_near_global_quarter_degree.jl:150]: I think this was a mistake

File: src/MultiRegion/multi_region_grid.jl
Comment (glwagner) [inline src/MultiRegion/multi_region_grid.jl:187]:
```suggestion
    global_ib = reconstruct_global_boundary(mrg.immersed_boundary)
    global_ib = on_architecture(architecture(mrg), global_ib)
```

when we leave the word "immersed" from "immersed boundary" things quickly get confusing

File: src/ImmersedBoundaries/partial_cell_bottom.jl
Comment (glwagner) [inline src/ImmersedBoundaries/partial_cell_bottom.jl:148]: This should go in a different PR.

File: src/TurbulenceClosures/TurbulenceClosures.jl
Comment (simone-silvestri) [reply src/TurbulenceClosures/TurbulenceClosures.jl:58]: it's the new rearrangement of the modules.

File: src/ImmersedBoundaries/partial_cell_bottom.jl
Comment (simone-silvestri) [reply src/ImmersedBoundaries/partial_cell_bottom.jl:148]: The new numerical bottom height is located at by convention at `z⁻ + Δz * (1 - ϵ)` so `ϵ * full_Δz` == `z - h` removing the necessity of the max. I thought this change can clean up the partial cell code

File: src/TurbulenceClosures/TurbulenceClosures.jl
Comment (simone-silvestri) [reply src/TurbulenceClosures/TurbulenceClosures.jl]: I have changed z_bottom to be only in `TurbulenceClosures` as it is the only module in which is needed / used and all similar methods (`z_top`, `depth`, `height_above_bottom`...) are defined only there

**Comment by simone-silvestri** [comment]
The dependency on OrthogonalSphericalShellGrids here is blocking the progress (the `SplitExplicitAuxiliaryFields` type has changed in this PR). How do we deal with this? We could remove the dependency in the tests of Oceananigans and move those tests over at OrthogonalSphericalShellGrids

File: src/ImmersedBoundaries/partial_cell_bottom.jl
Comment (simone-silvestri) [reply src/ImmersedBoundaries/partial_cell_bottom.jl:148]: actually, there is a bug here with how the immersed bottom height is computed. I ll correct it

**Comment by simone-silvestri** [comment]
> The dependency on OrthogonalSphericalShellGrids here is blocking the progress (the `SplitExplicitAuxiliaryFields` type has changed in this PR). How do we deal with this? We could remove the dependency in the tests of Oceananigans and move those tests over at OrthogonalSphericalShellGrids

Shall we do this? In case we want to do it I have prepared https://github.com/CliMA/OrthogonalSphericalShellGrids.jl/pull/46.
If yes, we can merge that one then this one, then update OrthogonalSphericalShellGrids with the new interface

**Comment by simone-silvestri** [comment]
This PR should be ready to merge

**Comment by simone-silvestri** [comment]
This PR requires an approval

File: src/ImmersedBoundaries/ImmersedBoundaries.jl
Comment (glwagner) [inline src/ImmersedBoundaries/ImmersedBoundaries.jl:97]: isn't this still useful? Why delete it?

File: src/ImmersedBoundaries/grid_fitted_bottom.jl
Comment (glwagner) [inline src/ImmersedBoundaries/grid_fitted_bottom.jl:146]: when this happens it could suggest that this should be more like `node` with

```julia
static_column_depth(i, j, grid, lx, ly)
```

File: src/ImmersedBoundaries/partial_cell_bottom.jl
Comment (glwagner) [reply src/ImmersedBoundaries/partial_cell_bottom.jl:165]: i just dont understand how this is related to a PR that implements a new function called `static_column_depth`

**Comment by glwagner** [comment]
I'm confused about this PR. There are all sorts of changes in different places. There is even a change to the Makie extension. What am I reviewing?

File: src/ImmersedBoundaries/ImmersedBoundaries.jl
Comment (simone-silvestri) [reply src/ImmersedBoundaries/ImmersedBoundaries.jl:97]: I have moved it in the `TurbulenceClosures` module where also `z_top` and `height_above_bottom` are defined

**Comment by simone-silvestri** [comment]
I have merged in #3880 because I wanted to see if the GPU tests pass. I am planning to merge this PR after that one. The rest is all from this PR. The OceananigansMakieExt change is a typo fix, but I see you have added a PR for it.

File: src/ImmersedBoundaries/ImmersedBoundaries.jl
Comment (glwagner) [reply src/ImmersedBoundaries/ImmersedBoundaries.jl:97]: Ok so its unrelated to the static column depth? I feel z_bottom belongs in the Grids module. What do you think?

Comment (glwagner) [reply src/ImmersedBoundaries/ImmersedBoundaries.jl:97]: Consider that it has to be extended for different immersed boundary implementation so it depends on partial cell bottom, etc. 

The original implementation may not have been optimal, but since we are moving things around here, I think we should try to improve on the original organization.

File: src/TurbulenceClosures/TurbulenceClosures.jl
Comment (glwagner) [inline src/TurbulenceClosures/TurbulenceClosures.jl:138]: is this used?

Comment (simone-silvestri) [reply src/TurbulenceClosures/TurbulenceClosures.jl:138]: seems to be used in CATKE here for example https://github.com/CliMA/Oceananigans.jl/blob/6c40d7e225c2127051b2703b9c62a8b18260e3a5/src/TurbulenceClosures/turbulence_closure_implementations/TKEBasedVerticalDiffusivities/catke_mixing_length.jl#L62

File: src/ImmersedBoundaries/ImmersedBoundaries.jl
Comment (simone-silvestri) [reply src/ImmersedBoundaries/ImmersedBoundaries.jl:97]: I am ok with moving all these functions in the Grids module, shall we do it in this PR?

File: src/ImmersedBoundaries/partial_cell_bottom.jl
Comment (glwagner) [inline src/ImmersedBoundaries/partial_cell_bottom.jl:94]:
```suggestion
        partial_Δz = z⁺ - zb
        too_thin = partial_Δz < ϵ * Δz
        numerical_zb  = ifelse(too_thin, zb, z⁺)
```

File: src/ImmersedBoundaries/partial_cell_bottom.jl
Comment (glwagner) [inline src/ImmersedBoundaries/partial_cell_bottom.jl:97]:
```suggestion
    end
    @inbounds bottom_field[i, j, 1] = numerical_zb
```

Comment (glwagner) [reply src/ImmersedBoundaries/partial_cell_bottom.jl:97]: should be outside the `k`-loop

File: src/ImmersedBoundaries/partial_cell_bottom.jl
Comment (glwagner) [reply src/ImmersedBoundaries/partial_cell_bottom.jl:93]: Which is a fine proposal but I think you should change the title of the PR. Imagine us looking for what is causing this kind of different in the future, combing through PRs. It's pretty hidden in this one

File: src/ImmersedBoundaries/partial_cell_bottom.jl
Comment (glwagner) [reply src/ImmersedBoundaries/partial_cell_bottom.jl:94]: this just makes it easier to understand. I was having trouble with the math. Also I think it's convoluted to evaluate the criterion by using the interface _below_ and then adding `Δz` to both sides, isn't it?

File: src/ImmersedBoundaries/partial_cell_bottom.jl
Comment (glwagner) [reply src/ImmersedBoundaries/partial_cell_bottom.jl:97]: and there's no need for an if statement

File: src/ImmersedBoundaries/partial_cell_bottom.jl
Comment (simone-silvestri) [reply src/ImmersedBoundaries/partial_cell_bottom.jl:93]: Hmmm, I thought there was supposed to be no difference. 
Here, the `bottom_height` is set to the actual bottom height if it is smaller than `z⁻ + ϵ Δz`, otherwise the cell is closed. 
So, the only difference with main will be for bottom heights that are in between `z⁻ + ϵ Δz` and `z⁺` (where `z⁻` and `z⁺` are the interfaces of the bottom cell.
I will check the difference in the internal_tide example

Comment (glwagner) [reply src/ImmersedBoundaries/partial_cell_bottom.jl:93]: But I thought before we would keep the cell open (rather than closing it).

Actually I think that MITgcm uses some criteria to decide which one to do --- either snapping up or down depending on which change is least.

Comment (simone-silvestri) [reply src/ImmersedBoundaries/partial_cell_bottom.jl:93]: ah, ok, I understand; open till `ϵ Δz` instead of snapping up.

Comment (glwagner) [reply src/ImmersedBoundaries/partial_cell_bottom.jl:93]: Actually I think we used to use that criteria too. It's something like 

```julia
partial_Δz = z⁺ - zb
too_thin = partial_Δz < ϵ * Δz
snap_up = partial_Δz < ϵ * Δz / 2
snapped_zb = ifelse(snap_up, z⁺, z⁻ + ϵ * Δz)
numerical_zb  = ifelse(too_thin, snapped_zb, zb)
```

**Comment by simone-silvestri** [comment]
In the new commit, I reverted to the previous implementation, so now there is no difference

https://github.com/user-attachments/assets/761fd5b0-e54c-40f6-b924-ab23ed69c17f

**Review by glwagner — APPROVED** [review]

File: src/ImmersedBoundaries/partial_cell_bottom.jl
Comment (simone-silvestri) [reply src/ImmersedBoundaries/partial_cell_bottom.jl]: right, I have modified it

**Pull request merged**
